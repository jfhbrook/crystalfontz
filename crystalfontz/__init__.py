# -*- coding: utf-8 -*-

"""
This is the documentation for crystalfontz.
"""

