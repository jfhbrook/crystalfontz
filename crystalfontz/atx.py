from dataclasses import dataclass
from enum import Enum
from typing import Iterable, Optional, Self

AUTO_POLARITY = 0x01


class AtxPowerSwitchFunction(Enum):
    RESET_INVERT = 0x02
    POWER_INVERT = 0x04
    LCD_OFF_IF_HOST_IS_OFF = 0x10
    KEYPAD_RESET = 0x20
    KEYPAD_POWER_ON = 0x40
    KEYPAD_POWER_OFF = 0x80


@dataclass
class AtxPowerSwitchFunctionalitySettings:
    functions: Iterable[AtxPowerSwitchFunction]
    auto_polarity: bool = True
    power_pulse_length_seconds: Optional[float] = None

    def to_bytes(self: Self) -> bytes:
        functions: int = 0
        for function in self.functions:
            functions ^= function.value
        if self.auto_polarity:
            functions ^= AUTO_POLARITY

        packed: bytes = functions.to_bytes()

        if self.power_pulse_length_seconds is not None:
            pulse_length = int(self.power_pulse_length_seconds * 32)

            if pulse_length < 1:
                raise ValueError(f"Pulse length can not be less than {1/32}")

            packed += min(pulse_length, 255).to_bytes()

        return packed
