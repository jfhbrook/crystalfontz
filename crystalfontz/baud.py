from typing import Literal

BaudRate = Literal[19200] | Literal[115200]
