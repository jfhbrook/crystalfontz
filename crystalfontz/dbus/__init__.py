try:
    from crystalfontz.dbus.client import DbusClient
    from crystalfontz.dbus.interface import DBUS_NAME, DbusInterface
except ImportError:
    DbusClient = None
    DBUS_NAME = None
    DbusInterface = None

__all__ = [
    "DbusClient",
    "DbusInterface",
    "DBUS_NAME",
]
