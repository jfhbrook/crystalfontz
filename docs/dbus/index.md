# DBus Support

The `crystalfontz` library includes a DBus interface, service and client. This service allows for multitenancy on Linux - the centralized service controls the serial bus, and clients (including `crystalfontzctl`) can connect to the service.

For information on the API, visit the API docs for [`crystalfontz.dbus.interface`](../api/crystalfontz.dbus.interface.md) and [`crystalfontz.dbus.client`](../api/crystalfontz.dbus.client.md).
