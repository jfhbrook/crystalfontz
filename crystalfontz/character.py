from typing import Dict, Self

from crystalfontz.error import EncodeError

"""
Abstractions for representing character ROMs (CGROMs), and encoding text to be
used with them.
"""

# Characters that take more than one code point in unicode. These need to be
# manually added to the character ROM.
super_minus: str = "\u207b"
super_one: str = "\u00B9"
inverse: str = f"{super_minus}{super_one}"
x_bar: str = "x̄"

# Characters which only require one code point, but are difficult to type
# and/or are ambiguous.
hbar = "―"  # Intended to be a katakana character?
block = "█"

# Japanese punctuation. The CFA533, at least, contains Katakana characters
# with corresponding punctuation.
japan_interpunct = "・"
japan_lquote = "「"
japan_rquote = "」"
japan_full_stop = "。"
japan_comma = "、"


class CgRom:
    def __init__(self: Self, sheet: str) -> None:
        self._table: Dict[str, bytes] = dict()

        lines = sheet.split("\n")
        if lines[0] == "":
            lines = lines[1:]
        if lines[-1] == "":
            lines = lines[0:-1]

        for i, row in enumerate(lines):
            for j, char in enumerate(row):
                point = (16 * j) + i
                if char != " " or point == 32:
                    self._table[char] = point.to_bytes()

    def __getitem__(self: Self, key: str) -> bytes:
        return self._table[key]

    def __setitem__(self: Self, key: str, value: bytes) -> None:
        self._table[key] = value

    def set_encoding(self: Self, char: str, encoded: int | bytes) -> Self:
        if isinstance(encoded, int):
            self[char] = encoded.to_bytes()
        else:
            self[char] = encoded
        return self

    def encode(self: Self, input: str, errors="strict") -> bytes:
        output: bytes = b""
        i = 0
        while i < len(input):
            char = input[i]

            # TODO: This encoder uses if/else statements to handle multi-byte
            # encodings. To make this general purpose, it needs to be
            # refactored to use a trie-like structure.
            if char == "x":
                if input[i + 1] == x_bar[1] and input[i + 2] == x_bar[2]:
                    output += self._table[x_bar]
                    i += 2
                else:
                    output += self._table["x"]
            elif char in self._table:
                output += self._table[char]
            elif char == super_minus:
                if input[i + 1] == super_one:
                    output += self._table[inverse]
                    i += 1
            else:
                if errors == "strict":
                    raise EncodeError(f"Unknown character {char}")
                else:
                    output += self._table["*"]

            i += 1

        return output


#
# This ROM encoding is based on page 44 of this doc:
#
#     file:///Users/josh/Downloads/CFA533-TMI-KU.pdf
#
# However, it is *incomplete*, mostly because I don't know katakana and only
# know a smattering of Greek. Unknown characters are filled in with spaces.
# Some characters that *are* filled out are best guesses.
#
# NOTE: ASCII characters generally share their code points with true ASCII.
# TODO: Does this ROM match another encoding which contains both katakana and
# Greek letters?
# NOTE: The first column in the ROM is reserved for custom characters.
#

ROM = (
    CgRom(
        """
   0@P`p   ―  α 
  !1AQaq  。ア  ä 
  "2BRbr  「 ツ βθ
  #3CScs  」ウ  ε∞
  $4DTdt  、エ  μΩ
  %5EUeu  ・オ  σü
  &6FVfv  ヲカ  ρΣ
  '7GWgw  アキ   π
  (8HXhx  イク  √ 
  )9IYiy  ゥケ    
  *:JZjz  エコ     
  +;K[k{  オサ    
  ,<L¥l|  ヤシ  ¢ 
  -=M]m}  ユヌ  £÷
  .>N^n→  ヨセ  ñ 
  /?O_o←  ツソ °ö█
"""
    )
    .set_encoding(inverse, 244 + 9)
    .set_encoding(x_bar, 240 + 8)
)


# TODO: This encoding is likely device specific. I'm leaving it in here for
# now because the encode method contains logic specific to the device. However,
# once that is factored out in favor of a trie-like data structure, this
# specific instance should be moved to device.py.
encode_chars = ROM.encode
