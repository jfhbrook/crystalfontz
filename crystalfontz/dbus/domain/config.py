from typing import Any, cast, ClassVar, Optional, Self, Tuple, Type

from crystalfontz.config import Config
from crystalfontz.dbus.domain.base import (
    ModelM,
    ModelT,
    RetryTimesM,
    RetryTimesT,
    RevisionM,
    RevisionT,
    struct,
    TimeoutM,
    TimeoutT,
)
from crystalfontz.dbus.domain.baud import BaudRateM, BaudRateT

FileT = str


class FileM:
    t: ClassVar[str] = "s"
    none: ClassVar[str] = ""

    @classmethod
    def pack(cls: Type[Self], file: Optional[str]) -> str:
        return file or cls.none


PortT = str


class PortM:
    t: ClassVar[str] = "s"


ConfigT = Tuple[
    FileT, PortT, ModelT, RevisionT, RevisionT, BaudRateT, TimeoutT, RetryTimesT
]


class ConfigM:
    """
    Map a config to and from dbus types.
    """

    t: ClassVar[str] = struct(
        FileM,
        PortM,
        ModelM,
        RevisionM,
        RevisionM,
        BaudRateM,
        TimeoutM,
        RetryTimesM,
    )

    @staticmethod
    def pack(config: Config) -> ConfigT:
        return (
            FileM.pack(config.file),
            config.port,
            config.model,
            RevisionM.pack(config.hardware_rev),
            RevisionM.pack(config.firmware_rev),
            config.baud_rate,
            TimeoutM.pack(config.timeout),
            RetryTimesM.pack(config.retry_times),
        )

    @staticmethod
    def unpack(config: ConfigT) -> Config:
        (
            file,
            port,
            model,
            hardware_rev,
            firmware_rev,
            baud_rate,
            timeout,
            retry_times,
        ) = config

        return cast(Any, Config)(
            file=file,
            port=port,
            model=model,
            hardware_rev=RevisionM.unpack(hardware_rev),
            firmware_rev=RevisionM.unpack(firmware_rev),
            baud_rate=baud_rate,
            timeout=TimeoutM.unpack(timeout),
            retry_times=RetryTimesM.unpack(retry_times),
        )
