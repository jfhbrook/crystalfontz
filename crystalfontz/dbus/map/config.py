from typing import Any, cast, ClassVar, Optional, Self, Type

from crystalfontz.config import Config
from crystalfontz.dbus.config import ConfigStruct
from crystalfontz.dbus.map.base import RetryTimesM, struct, TimeoutM
from crystalfontz.dbus.map.baud import BaudRateM


class ConfigFileM:
    t: ClassVar[str] = "s"
    none: ClassVar[str] = ""

    @classmethod
    def pack(cls: Type[Self], file: Optional[str]) -> str:
        return file or cls.none


class PortM:
    t: ClassVar[str] = "s"


class ModelM:
    t: ClassVar[str] = "s"


class RevisionM:
    t: ClassVar[str] = "s"
    none: ClassVar[str] = ""

    @classmethod
    def pack(cls: Type[Self], revision: Optional[str]) -> str:
        return revision or cls.none

    @classmethod
    def unpack(cls: Type[Self], revision: str) -> Optional[str]:
        return revision if revision != cls.none else None


class ConfigM:
    t: ClassVar[str] = struct(
        ConfigFileM,
        PortM,
        ModelM,
        RevisionM,
        RevisionM,
        BaudRateM,
        TimeoutM,
        RetryTimesM,
    )

    @staticmethod
    def pack(config: Config) -> ConfigStruct:
        return (
            ConfigFileM.pack(config.file),
            config.port,
            config.model,
            RevisionM.pack(config.hardware_rev),
            RevisionM.pack(config.firmware_rev),
            config.baud_rate,
            TimeoutM.pack(config.timeout),
            RetryTimesM.pack(config.retry_times),
        )

    @staticmethod
    def unpack(config: ConfigStruct) -> Config:
        (
            file,
            port,
            model,
            hardware_rev,
            firmware_rev,
            baud_rate,
            timeout,
            retry_times,
        ) = config

        return cast(Any, Config)(
            file=file,
            port=port,
            model=model,
            hardware_rev=RevisionM.unpack(hardware_rev),
            firmware_rev=RevisionM.unpack(firmware_rev),
            baud_rate=baud_rate,
            timeout=TimeoutM.unpack(timeout),
            retry_times=RetryTimesM.unpack(retry_times),
        )
