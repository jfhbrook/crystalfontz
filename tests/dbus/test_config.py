from typing import Any, cast

import pytest

from crystalfontz.baud import FAST_BAUD_RATE, SLOW_BAUD_RATE
from crystalfontz.config import Config, GLOBAL_FILE

try:
    from crystalfontz.dbus.config import StagedConfig
except ImportError:
    StagedConfig = None

cfg_cls = cast(Any, Config)


@pytest.mark.parametrize(
    "active_config,target_config",
    [
        (
            cfg_cls(
                file=GLOBAL_FILE,
                port="/dev/ttyS0",
                model="CFA533",
                hardware_rev="h1.4",
                firmware_rev="u1v2",
                baud_rate=FAST_BAUD_RATE,
                timeout=0.250,
                retry_times=1,
            ),
            cfg_cls(
                file=GLOBAL_FILE,
                port="/dev/ttyS0",
                model="CFA533",
                hardware_rev="h1.4",
                firmware_rev="u1v2",
                baud_rate=FAST_BAUD_RATE,
                timeout=0.250,
                retry_times=1,
            ),
        ),
        (
            cfg_cls(
                file=GLOBAL_FILE,
                port="/dev/ttyS0",
                model="CFA533",
                hardware_rev="h1.4",
                firmware_rev="u1v2",
                baud_rate=FAST_BAUD_RATE,
                timeout=0.250,
                retry_times=1,
            ),
            cfg_cls(
                file=GLOBAL_FILE,
                port="/dev/ttyS4",
                model="CFA533",
                hardware_rev="h1.4",
                firmware_rev="u1v2",
                baud_rate=SLOW_BAUD_RATE,
                timeout=0.250,
                retry_times=1,
            ),
        ),
    ],
)
def test_staged_config_as_dict(active_config, target_config, snapshot) -> None:
    cls = cast(Any, StagedConfig)
    staged = cls(active_config=active_config, target_config=target_config)
    assert staged.as_dict() == snapshot


@pytest.mark.parametrize(
    "active_config,target_config",
    [
        (
            cfg_cls(
                file=GLOBAL_FILE,
                port="/dev/ttyS0",
                model="CFA533",
                hardware_rev="h1.4",
                firmware_rev="u1v2",
                baud_rate=FAST_BAUD_RATE,
                timeout=0.250,
                retry_times=1,
            ),
            cfg_cls(
                file=GLOBAL_FILE,
                port="/dev/ttyS0",
                model="CFA533",
                hardware_rev="h1.4",
                firmware_rev="u1v2",
                baud_rate=FAST_BAUD_RATE,
                timeout=0.250,
                retry_times=1,
            ),
        ),
        (
            cfg_cls(
                file=GLOBAL_FILE,
                port="/dev/ttyS0",
                model="CFA533",
                hardware_rev="h1.4",
                firmware_rev="u1v2",
                baud_rate=FAST_BAUD_RATE,
                timeout=0.250,
                retry_times=1,
            ),
            cfg_cls(
                file=GLOBAL_FILE,
                port="/dev/ttyS4",
                model="CFA533",
                hardware_rev="h1.4",
                firmware_rev="u1v2",
                baud_rate=SLOW_BAUD_RATE,
                timeout=0.250,
                retry_times=1,
            ),
        ),
    ],
)
def test_staged_config_repr(active_config, target_config, snapshot) -> None:
    cls = cast(Any, StagedConfig)
    staged = cls(active_config=active_config, target_config=target_config)
    assert repr(staged) == snapshot
