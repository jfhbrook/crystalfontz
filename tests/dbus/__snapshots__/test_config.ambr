# serializer version: 1
# name: test_staged_config_as_dict[active_config0-target_config0]
  dict({
    'baud_rate': dict({
      'active': 115200,
      'target': 115200,
      'type': None,
    }),
    'file': dict({
      'active': '/etc/crystalfontz.yaml',
      'target': '/etc/crystalfontz.yaml',
      'type': None,
    }),
    'firmware_rev': dict({
      'active': 'u1v2',
      'target': 'u1v2',
      'type': None,
    }),
    'hardware_rev': dict({
      'active': 'h1.4',
      'target': 'h1.4',
      'type': None,
    }),
    'model': dict({
      'active': 'CFA533',
      'target': 'CFA533',
      'type': None,
    }),
    'port': dict({
      'active': '/dev/ttyS0',
      'target': '/dev/ttyS0',
      'type': None,
    }),
    'retry_times': dict({
      'active': 1,
      'target': 1,
      'type': None,
    }),
    'timeout': dict({
      'active': 0.25,
      'target': 0.25,
      'type': None,
    }),
  })
# ---
# name: test_staged_config_as_dict[active_config1-target_config1]
  dict({
    'baud_rate': dict({
      'active': 115200,
      'target': 19200,
      'type': 'set',
    }),
    'file': dict({
      'active': '/etc/crystalfontz.yaml',
      'target': '/etc/crystalfontz.yaml',
      'type': None,
    }),
    'firmware_rev': dict({
      'active': 'u1v2',
      'target': 'u1v2',
      'type': None,
    }),
    'hardware_rev': dict({
      'active': 'h1.4',
      'target': 'h1.4',
      'type': None,
    }),
    'model': dict({
      'active': 'CFA533',
      'target': 'CFA533',
      'type': None,
    }),
    'port': dict({
      'active': '/dev/ttyS0',
      'target': '/dev/ttyS4',
      'type': 'set',
    }),
    'retry_times': dict({
      'active': 1,
      'target': 1,
      'type': None,
    }),
    'timeout': dict({
      'active': 0.25,
      'target': 0.25,
      'type': None,
    }),
  })
# ---
# name: test_staged_config_repr[active_config0-target_config0]
  '''
    baud_rate: '115200'
    file: /etc/crystalfontz.yaml
    firmware_rev: u1v2
    hardware_rev: h1.4
    model: CFA533
    port: /dev/ttyS0
    retry_times: '1'
    timeout: '0.25'
    
  '''
# ---
# name: test_staged_config_repr[active_config1-target_config1]
  '''
  ~ baud_rate: 115200 ~> 19200
    file: /etc/crystalfontz.yaml
    firmware_rev: u1v2
    hardware_rev: h1.4
    model: CFA533
  ~ port: /dev/ttyS0 ~> /dev/ttyS4
    retry_times: '1'
    timeout: '0.25'
    
  '''
# ---
